genie-agent/src/main/java/com/netflix/genie/agent/execution/statemachine/actions/LaunchJobAction.java

```diff
@@ -35,13 +40,55 @@
 @Lazy
 class LaunchJobAction extends BaseStateAction implements StateAction.LaunchJob {
 
+    private final ExecutionContext executionContext;
+    private final LaunchJobService launchJobService;
+
+    LaunchJobAction(
+        final ExecutionContext executionContext,
+        final LaunchJobService launchJobService
+    ) {
+        this.executionContext = executionContext;
+        this.launchJobService = launchJobService;
+    }
+
     /**
      * {@inheritDoc}
      */
     @Override
     protected Events executeStateAction(final StateContext<States, Events> context) {
         log.info("Launching job...");
-        //TODO implement this action
+
+        final Process jobProcess;
+        try {
+            jobProcess = launchJobService.launchProcess();
+        } catch (final JobLaunchException e) {
+            throw new RuntimeException("Failed to launch job", e);
+        }
+
+        log.info("Job process started (pid: {})", getPid(jobProcess));
+
+        executionContext.setJobProcess(jobProcess);
+
         return Events.LAUNCH_JOB_COMPLETE;
     }
+
+    /* TODO: HACK, Process does not expose PID in Java 8 API */
```

Comment:
Since this isn't available in Java 8 what is the TODO here? Just to find a better way?

Reply:
I have to do this hack because Java 8 does not expose PID.
Java 9 does, so in the future we may clean this up and just call `getPid()`

Reply:
Ok so it will stay as is for now until we can move to Java 9+ (probably Java 11 as I believe it's the next LTS version)